LFS-242: As an administrator, I can configure sections/questions within a questionnaire to be displayed conditionally based on the answers provided by the user to previous questions within that questionnaire

OrdiNeu opened the pull request:
[Jira link](https://phenotips.atlassian.net/browse/LFS-242).

**Testing branch**: `LFS-242-test`. In demographics:
- one of the sections will only appear if `patient_id` is set to 5.
- one of the sections will only appear If `patient_id` is 6 OR if the family is set to 'test'
- one of the sections will only appear if `patient_id` is 6 AND if the family is set to 'test'

This commit enables the concept of conditional sections, where a value will only appear if other values defined elsewhere in the form meets some criteria.

OrdiNeu committed:
As an administrator, I can configure sections/questions within a
questionnaire to be displayed conditionally based on the answers
provided by the user to previous questions within that questionnaire

Add basic conditionals

OrdiNeu committed:
LFS-242: As an administrator, I can configure sections/questions within
a questionnaire to be displayed conditionally based on the answers
provided by the user to previous questions within that questionnaire

Fix possible infinite loop. Code cleanup

OrdiNeu committed:
LFS-242: As an administrator, I can configure sections/questions within
a questionnaire to be displayed conditionally based on the answers
provided by the user to previous questions within that questionnaire

Add support for conditional groups.

OrdiNeu committed:
LFS-242: As an administrator, I can configure sections/questions within
a questionnaire to be displayed conditionally based on the answers
provided by the user to previous questions within that questionnaire

Refactor conditionals to use a component manager.

OrdiNeu committed:
LFS-242: As an administrator, I can configure sections/questions within
a questionnaire to be displayed conditionally based on the answers
provided by the user to previous questions within that questionnaire

Code style improvements.

veronikaslc commented:
Tested.

OrdiNeu commented:
Of note: without [LFS-209](https://phenotips.atlassian.net/browse/LFS-209), we can't implement cross-form-content-linking (e.g. display section only if the person's TP53 status, _as entered in another form_, is <x>).

marta- reviewed (changes requested):
Data entered in conditionally displayed sections that are hidden at the moment of saving ends up saved in the database, and this should not happen.

Steps to reproduce on the test branch:
- open a demographics form
- enter 5 in the patient id -> the question "How is patient 5 feeling today?" will appear
- enter "Fine" as an answer to that question
- go back to patient id and enter a different value -> the question "How is patient 5 feeling today?" will be hidden
- SAVE
- reopen/refresh the form
- change patient id back to 5 -> the question  "How is patient 5 feeling today?" **with the answer "fine"** will appear.

This is not right because we're storing irrelevant data that will skew the search/filtering. For example, in a more realistic scenario, we might want to filter by a certain chemotherapy agent. The results could include a patient for whom the answer to "Has undergone chemo" is "No" but has some leftover data entered by mistake.

To clarify, the data should be preserved in the form while the user is editing the form, but should not be stored in the database and should not appear when the user reopens the form.

OrdiNeu committed:
LFS-242: As an administrator, I can configure sections/questions within
a questionnaire to be displayed conditionally based on the answers
provided by the user to previous questions within that questionnaire

Unmount invisible sections so that they are not saved.

OrdiNeu commented:
Point addressed by unmounting the DOM node while it's hidden, but I found another bug where it will refuse to update the content of a section if the answer is empty. I'm investigating it now.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/ConditionalSingle.jsx:
```diff
@@ -0,0 +1,92 @@
+    // Invalid operand
+    throw new Error("Invalid operand specified.")
+  }
+}
```
Such a beautiful conditional manager architecture above, only to have IFs here... :)

Suggestion:
Declare a map from operator to comparison function:

```
var COMPARE_MAP = {
 '=' : (a, b) => {return a == b},
 '<' : (a, b) => {return a < b},
...
'is empty' : (a) => {return a === null || a === undefined},
...
};
```
then rewrite this function as:
```
export function isConditionalSatisfied(comparator, ...operands) {
  return COMPARE_MAP[comparator]?.apply(null, operands);
}
```
This allows for more flexibility in terms of cumber of operands, and makes it somewhat easier to add support for new operators.

OrdiNeu replied:
Refactored, although since I wanted more verbose error handling I've also inserted a guard against the operand not being inside COMPARE_MAP.

OrdiNeu committed:
LFS-242: As an administrator, I can configure sections/questions within
a questionnaire to be displayed conditionally based on the answers
provided by the user to previous questions within that questionnaire

Refactor the conditional single logic.

marta- reviewed (approved):
Looks good. I can see opportunities for additional refactoring to make the code even more maintainable/extensible, but not in the scope of this sprint.
